### PR TITLE
Allow explicit empty security definition to overwrite existing definitions

### PR DESCRIPTION
**Pull request by co3k**
It enables to clear default security configuration in operationObject.Security (to make public api).

protoc-gen-swagger doesn't allow empty array (and it causes segfault) in operationObject.Security but it is really needed resetting security requirement configuration by Swagger.

### CODE REVIEW

**Comment by co3k** [comment]
I wonder why the https://travis-ci.org/grpc-ecosystem/grpc-gateway/builds/390065893 build fails in   `Go: 1.10.x, GATEWAY_PLUGIN_FLAGS=` job. Can anyone advice me to solve this problem?

**Comment by yugui** [comment]
The test failure is because the generated files in `examples/` directory are inconsistent to the change.
You can fix it by running `make examples` at the top level of the repository and confirming that the diff meets your expectation.

**Comment by yugui** [comment]
The test failure looks to be because some of example files are out-of-date. c.f. https://github.com/golang/protobuf/commit/05f48f4eaf0e05663b562bab533cdd472238ce29#diff-8c603013608023320d5242916c4ea03bR1571

Would you mind running `make clean && make examples` to let the files catch the change in protoc-gen-go?

**Comment by co3k** [comment]
@yugui
Thank you for your response, but I can't get any diffs by running that command on my macOS.

```
$ go version
go version go1.10.3 darwin/amd64

$ protoc --version
libprotoc 3.5.1

$ brew info protobuf
protobuf: stable 3.5.1 (bottled), HEAD
Protocol buffers (Google's data interchange format)
https://github.com/google/protobuf/
/usr/local/Cellar/protobuf/3.2.0_1 (257 files, 15.9MB)
  Poured from bottle on 2017-04-28 at 18:02:34
/usr/local/Cellar/protobuf/3.5.1_1 (267 files, 18.5MB) *
  Poured from bottle on 2018-06-11 at 00:22:59
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/protobuf.rb
==> Dependencies
Build: autoconf ✔, automake ✔, libtool ✔
Recommended: python@2 ✔
Optional: python ✘
==> Options
--with-python
	Build with python support
--with-test
	Run build-time check
--without-python@2
	Build without python2 support
--HEAD
	Install HEAD version
==> Caveats
Editor support and examples have been installed to:
  /usr/local/opt/protobuf/share/doc/protobuf

$ cd $GOPATH/src/github.com/golang/protobuf/protoc-gen-go; git log -1 --oneline; cd -
05f48f4 (HEAD -> master, origin/master, origin/HEAD) proto: revert UTF-8 validation for proto2 (#628)

$ make clean && make examples
rm -f bin/protoc-gen-grpc-gateway bin/protoc-gen-swagger
go build -o bin/protoc-gen-grpc-gateway github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway
**SNIP**

$ git status
On branch allow-empty-array-in-security
nothing to commit, working tree clean

$ make clean && make test
rm -f bin/protoc-gen-grpc-gateway bin/protoc-gen-swagger
go build -o bin/protoc-gen-grpc-gateway github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway
**SNIP**

$ git status             
On branch allow-empty-array-in-security
nothing to commit, working tree clean
```

**Comment by co3k** [comment]
The error has been fixed at https://github.com/grpc-ecosystem/grpc-gateway/pull/666/commits/d1985fcea6b588fc855bfa70b0600524ad1e4027 .
I might need to run `make realclean && make examples` for solving this problem.
Thanks again to yugui!

The https://travis-ci.org/grpc-ecosystem/grpc-gateway/jobs/390576866 job for `Go: 1.9.x, GATEWAY_PLUGIN_FLAGS=` has been timed out.
Is it a temporary error? But I cannot retry this job because I don't have permission to do so.

**Comment by tmc** [comment]
@co3k thanks for your contribution, can you rebase this so CI runs again?

**Comment by co3k** [comment]
@tmc Thanks for your response but the https://travis-ci.org/grpc-ecosystem/grpc-gateway/jobs/394549559#L554-L557 test is failed that is due to timed out in go get phase. I cannot rebase it again because the upstream is not updated yet. How should I do next?

**Comment by johanbrandhorst** [comment]
I have re-triggered the job.

File: protoc-gen-swagger/genswagger/types.go [hunk 1]
```diff
@@ -105,7 +105,7 @@ type swaggerOperationObject struct {
 	Tags        []string                `json:"tags,omitempty"`
 	Deprecated  bool                    `json:"deprecated,omitempty"`
 
-	Security     []swaggerSecurityRequirementObject  `json:"security,omitempty"`
+	Security     *[]swaggerSecurityRequirementObject `json:"security,omitempty"`
```
Comment (johanbrandhorst) [inline protoc-gen-swagger/genswagger/types.go]: Is this required? Can't we just have an empty slice?

File: protoc-gen-swagger/genswagger/template.go [hunk 2]
```diff
@@ -653,21 +653,26 @@ func renderServices(services []*descriptor.Service, paths swaggerPathsObject, re
 					}
 					if opts.Security != nil {
 						newSecurity := []swaggerSecurityRequirementObject{}
-						if operationObject.Security == nil {
-							newSecurity = []swaggerSecurityRequirementObject{}
-						} else {
-							newSecurity = operationObject.Security
+						if operationObject.Security != nil {
+							newSecurity = *operationObject.Security
 						}
```
Comment (johanbrandhorst) [inline protoc-gen-swagger/genswagger/template.go]: Can we not just set `newSecurity = nil` when `operationObject.Security == nil`?

File: protoc-gen-swagger/genswagger/types.go (hunk 1 above)
Comment (co3k) [reply protoc-gen-swagger/genswagger/types.go]: Required. The `omitempt` tag removes an empty slice. However we don't want to print `null` value in every JSON response so we can't remove this `omitempty` tag.

File: protoc-gen-swagger/genswagger/template.go (hunk 2 above)
Comment (co3k) [reply protoc-gen-swagger/genswagger/template.go]: I think we really need empty slice in this case.

File: protoc-gen-swagger/genswagger/types.go (hunk 1 above)
Comment (johanbrandhorst) [reply protoc-gen-swagger/genswagger/types.go]: OK

File: protoc-gen-swagger/genswagger/template.go (hunk 2 above)
Comment (johanbrandhorst) [reply protoc-gen-swagger/genswagger/template.go]: OK

**Comment by johanbrandhorst** [comment]
As with the other PR, could you please squash the commits? Thank you.

File: examples/proto/examplepb/a_bit_of_everything.proto [hunk 3]
```diff
@@ -274,6 +274,8 @@ service ABitOfEverythingService {
 				url: "https://github.com/grpc-ecosystem/grpc-gateway";
 				description: "Find out more about GetQuery";
 			}
+		  security: {
+      }
```
Comment (johanbrandhorst) [inline examples/proto/examplepb/a_bit_of_everything.proto]: The indendation of this line seems off, are you using spaces instead of tabs?

**Comment by johanbrandhorst** [comment]
Could you also please reword this title, description and commit message to clearly state the intent please? Something along the lines of:

_Allow explicit empty security definition to overwrite existing definitions._

File: examples/proto/examplepb/a_bit_of_everything.proto (hunk 3 above)
Comment (co3k) [reply examples/proto/examplepb/a_bit_of_everything.proto]: Sorry, I've fixed it in https://github.com/grpc-ecosystem/grpc-gateway/pull/666/commits/a78ff380159daf7f0da4009c5b65a2263ad32249 (I'll squash this commit later)

**Comment by co3k** [comment]
@johanbrandhorst Thanks for your advice! I've squashed commits and changed the title of this pull request and the first commit to follow your suggestion.

**Comment by johanbrandhorst** [comment]
Thank you for your contribution!

**Pull request merged**